hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/TranslateToPublicAddressProvider.java

```diff
@@ -0,0 +1,129 @@
+
+    private boolean resolve(AddressProvider addressProvider, Collection<MemberInfo> members) {
+        if (!(addressProvider instanceof DefaultAddressProvider)) {
+            return false;
+        }
+
+        // Default value of DISCOVERY_SPI_PUBLIC_IP_ENABLED is `null` intentionally.
+        // if DISCOVERY_SPI_PUBLIC_IP_ENABLED is not set to true/false, we don't know the intention of the user,
+        // we will try to decide if we should use private/public address automatically int that case.
+        String publicIpEnabledProperty = properties.getString(ClientProperty.DISCOVERY_SPI_PUBLIC_IP_ENABLED);
+        if (publicIpEnabledProperty == null) {
+            if (members.isEmpty() || memberInternalAddressAsDefinedInClientConfig(members)) {
+                return false;
+            }
+
+            return membersReachableOnlyViaPublicAddress(members);
+        }
+        return properties.getBoolean(ClientProperty.DISCOVERY_SPI_PUBLIC_IP_ENABLED);
+    }
+
+    /**
+     * Checks if any member has its internal address as configured in ClientConfig.
+     * <p>
+     * If any member has its internal/private address the same as configured in ClientConfig, then it means that the client is
+     * able to connect to members via configured address. No need to use make any address translation.
+     */
+    boolean memberInternalAddressAsDefinedInClientConfig(Collection<MemberInfo> members) {
+        List<String> addresses = config.getNetworkConfig().getAddresses();
+        return members.stream()
+                .map(MemberInfo::getAddress)
+                .anyMatch(a -> addresses.contains(a.getHost())
+                        || addresses.contains(String.format("%s:%s", a.getHost(), a.getPort())));
+    }
+
+    /**
+     * Checks if members are reachable via public addresses, but not reachable via internal addresses.
+     * <p>
+     * We check only limited number of random members to reduce the slowdown of the startup.
+     */
+    private boolean membersReachableOnlyViaPublicAddress(Collection<MemberInfo> members) {
+        Iterator<MemberInfo> iter = members.iterator();
+        for (int i = 0; i < REACHABLE_CHECK_NUMBER; i++) {
+            if (!iter.hasNext()) {
+                iter = members.iterator();
+            }
+            MemberInfo member = iter.next();
+            Address publicAddress = member.getAddressMap().get(EndpointQualifier.resolve(ProtocolType.CLIENT, "public"));
+            if (publicAddress == null) {
+                return false;
+            }
+            Address internalAddress = member.getAddress();
+            if (isReachable(internalAddress, REACHABLE_ADDRESS_TIMEOUT)) {
+                return false;
+            }
+            if (!isReachable(publicAddress, NON_REACHABLE_ADDRESS_TIMEOUT)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private boolean isReachable(Address address, int timeoutMs) {
+        try (Socket s = new Socket()) {
```

Comment:
I didn't experiment with the code, but will `java.net.Socket` work in case of secure connections? I'd expect an exception in this case, but I may be wrong.

Reply:
Update. I did some experiments and this approach seems to be working with a EE member which has TLS enabled. The only minor issue is that the member logs some warnings on connection attempt, like the following one:
```
javax.net.ssl.SSLException: Remote socket closed during SSL/TLS handshake.  This is probably caused by a SSL/TLS authentication problem resulting in the remote side closing the socket.
```

Reply:
I was not sure and tried the following code. 
```
char[] passphrase = "123456".toCharArray();
        SSLContext ctx = SSLContext.getInstance("TLS");
        KeyManagerFactory kmf = KeyManagerFactory.getInstance("SunX509");
        KeyStore ks = KeyStore.getInstance("JKS");
        ks.load(new FileInputStream("/Users/sancar/projects/hazelcast-enterprise/hazelcast-enterprise/src/test/resources/com/hazelcast/nio/ssl/hazelcast.keystore"), passphrase);
        kmf.init(ks, passphrase);
        ctx.init(kmf.getKeyManagers(), null, null);
        SSLServerSocketFactory ssf = ctx.getServerSocketFactory();
        SSLServerSocket ss = (SSLServerSocket) ssf.createServerSocket(5701);
        ss.setNeedClientAuth(true);

        new Thread(() -> {
            while (true) {
                try {
                    Socket accept = ss.accept();
                    System.out.println("ACCEPTED" + accept);
                } catch (Exception e) {
                    e.printStackTrace();
                }

            }
        }).start();
        
        try (Socket s = new Socket()) {
            s.connect(new InetSocketAddress("127.0.0.1", 5701), 5000);
            System.out.println("CONNECTED");
        } catch (Exception e) {
            e.printStackTrace();
        }
```

Reply:
I already did an experiment with a IMDG EE member which has TLS enabled. See https://github.com/hazelcast/hazelcast/pull/17895#discussion_r534226708

Reply:
So the logs on the server side do not look nice. To avoid the logs, 
1. we may open an SSL connection and close it after the initial handshake. The downside is this heuristic will slowdown the startup more. 
I have tried to use SSL here but porting SSL connection and configuration does not seem simple to me. Maybe @pveentjer can handle it better than me? 
Most of the SSL related connections are not in this repo. And the ClientExtension seems to be for only channels. I could not figure out how to simply open a connection and closed it back after the initial handshake. 

2. We can lower the SSLException log level from WARNING to FINE when the initial PROTOCOL bytes are not sent. It would mean that if normal member/client connections will also be suppressed when the connection is closed before bytes are sent.

@jerrinot @mmedenjak Dear colleagues, can we get your ideas ? :) 
And to give more context, we are trying to decide if a member is reachable by opening a connection from the client-side to make a decision whether to use `internal` or `public` address of the member. 

Secondly, if we leave it like this and allow the Warning logs will be once in a lifetime of the client(and also when they switch cluster). But if you have hundreds of clients, we will see hundreds of logs in the server. 

Also, we discussed ICMP port 7 pings with @puzpuzpuz . It does not show if a member is there but shows if the machine is reachable. It might be enough but, it is not always enabled.

Reply:
An update after getting a valid point from @leszko 
We have decided to go with checking `config.getNetworkConfig().getSSLConfig().isEnabled()`
If it is enabled we will not run this heuristic to decide automatically. 

The reasoning is that if the user enables SSL, s/he already has lots of configs, it is not an out-of-the-box experience anymore. The user can put one more property, DISCOVERY_SPI_PUBLIC_IP_ENABLED true/false.

Reply:
I'm ok with the proposed solution, but we really need to document this fancy logic, otherwise users will get confused.

Reply:
One more note on this one. We should add more logging around this detection mechanism to simplify diagnostics in case of reported issues.

Reply:
I have added logs.